man/ejdb2.3
.TH "EJDB2" 3 "2020-04-01" "Man Page" "EJDB2"

.SH NAME
EJDB2 \- Embedded JSON Database engine

.SH DESCRIPTION
.PP
EJDB2 is an embeddable JSON database engine published under MIT license.

.SH "SEE ALSO"
.I http://ejdb.org
EJDB project offcial web site.


Changelog
ejdb2 (2.0.36) testing; urgency=medium

  * Implemented Swift language binding #267
  * Ported to iOS
  * Upgraded to iowow 1.3.34
  * Upgraded to facil.io 0.7.4
  * Improved ejdb2.h documentation

 -- Anton Adamansky <adamansky@gmail.com>  Mon, 06 Jan 2020 23:20:38 +0700

ejdb2 (2.0.35) testing; urgency=medium

  * FIXED: JQL apply ignores value set by placeholder (#269)

 -- Anton Adamansky <adamansky@gmail.com>  Tue, 10 Dec 2019 21:12:43 +0700

ejdb2 (2.0.34) testing; urgency=medium

  * Reduced library size by 400K: stripped off unused utf8proc data

 -- Anton Adamansky <adamansky@gmail.com>  Thu, 05 Dec 2019 12:01:59 +0700

ejdb2 (2.0.33) testing; urgency=medium

  * Added ejdb_merge_or_put atomic method #268
  * Added getOrNull to node/react bindings

 -- Anton Adamansky <adamansky@gmail.com>  Wed, 04 Dec 2019 17:34:01 +0700

ejdb2 (2.0.32) testing; urgency=medium

  * Upgraded to iowow v1.3.31

 -- Anton Adamansky <adamansky@gmail.com>  Wed, 04 Dec 2019 17:33:20 +0700

ejdb2 (2.0.31) testing; urgency=medium

  * Upgraded to iowow v1.3.30

 -- Anton Adamansky <adamansky@gmail.com>  Tue, 19 Nov 2019 20:40:29 +0700

ejdb2 (2.0.30) testing; urgency=medium

  * Upgraded to iowow v1.3.29

 -- Anton Adamansky <adamansky@gmail.com>  Fri, 15 Nov 2019 13:55:03 +0700

ejdb2 (2.0.29) testing; urgency=medium

  * Upgraded to iowow v1.3.27
  * React Native binding implemented (Android)

 -- Anton Adamansky <adamansky@gmail.com>  Mon, 04 Nov 2019 21:52:07 +0700

ejdb2 (2.0.28) testing; urgency=medium

  * Upgraded to facil.io v0.7.3
  * JBS server functions removed from ejdb2 target libs

 -- Anton Adamansky <adamansky@gmail.com>  Fri, 27 Sep 2019 18:26:57 +0700

ejdb2 (2.0.27) testing; urgency=medium

  * Upgraded to iowow v1.3.25 with critical fixes

 -- Anton Adamansky <adamansky@gmail.com>  Thu, 29 Aug 2019 12:46:41 +0700

ejdb2 (2.0.26) testing; urgency=medium

  * Upgraded to iowow v1.3.24

 -- Anton Adamansky <adamansky@gmail.com>  Thu, 22 Aug 2019 02:16:50 +0700

ejdb2 (2.0.25) testing; urgency=medium

  * ejdb2_dart: Fixed memory leak in dart binding
  * Upgraded to iowow 1.3.23

 -- Anton Adamansky <adamansky@gmail.com>  Sat, 17 Aug 2019 21:27:29 +0700

ejdb2 (2.0.24) testing; urgency=medium

  * Strip shared libraries for release builds.

 -- Anton Adamansky <adamansky@gmail.com>  Thu, 25 Jul 2019 21:09:00 +0700

ejdb2 (2.0.23) testing; urgency=medium

  * BUG: Fixed incorrect behaviour of `jbi_ftoa`

 -- Anton Adamansky <adamansky@gmail.com>  Fri, 19 Jul 2019 05:55:16 +0700

ejdb2 (2.0.22) testing; urgency=medium

  * Limited windows support without HTTP/Websocket network API (#237)
  * `static_assert` is  set to `_Static_assert` if not defined
  * All uses of `iwftoa` replaced by `jbi_ftoa`

 -- Anton Adamansky <adamansky@gmail.com>  Thu, 18 Jul 2019 19:27:26 +0700

ejdb2 (2.0.21) testing; urgency=medium

  * Added `ejdb_rename_collection` method  (#254)

 -- Anton Adamansky <adamansky@gmail.com>  Fri, 12 Jul 2019 12:39:32 +0700

ejdb2 (2.0.20) testing; urgency=medium

  * Node.js binding (#248)
  * Added git revision hash getter `ejdb_git_revision()` (ejdb2.h)
  * Better error reporting on failed memory allocations

 -- Anton Adamansky <adamansky@gmail.com>  Wed, 03 Jul 2019 19:36:46 +0700

ejdb2 (2.0.19) testing; urgency=medium

  * Upgraded to iowow v1.3.20 with critical fixes

 -- Anton Adamansky <adamansky@gmail.com>  Thu, 13 Jun 2019 22:04:27 +0700

ejdb2 (2.0.18) testing; urgency=medium

  * Limit one time file allocation step to 2G iowow v1.3.18
  * Added Docker image (#249)
  * Better qsort_t detection, build ok with `musl`

 -- Anton Adamansky <adamansky@gmail.com>  Wed, 12 Jun 2019 16:48:57 +0700

ejdb2 (2.0.17) testing; urgency=medium

  * Added `inverse` JQL query option.

 -- Anton Adamansky <adamansky@gmail.com>  Wed, 05 Jun 2019 23:15:09 +0700

ejdb2 (2.0.16) testing; urgency=medium

  * Allowed dash and underscope chars in collection names specified in query: `[@collection_name]/query`.
  * Fixed minor typo in ejdb2_jni/src/CMakeLists.txt

 -- Anton Adamansky <adamansky@gmail.com>  Wed, 05 Jun 2019 19:14:38 +0700

ejdb2 (2.0.15) testing; urgency=medium

  * Security: a heap-overflow vulnerability was fixed in the WebSocket parser of facil.io library

 -- Anton Adamansky <adamansky@gmail.com>  Mon, 27 May 2019 13:33:03 +0700

ejdb2 (2.0.14) testing; urgency=medium

  * Android support
  * Test cases for Android JNI binding

 -- Anton Adamansky <adamansky@gmail.com>  Fri, 17 May 2019 00:57:11 +0700
ejdb2 (2.0.13) testing; urgency=medium

  * Upgraded to iowow_1.3.17 with critical fixes

 -- Anton Adamansky <adamansky@gmail.com>  Fri, 03 May 2019 18:27:10 +0700

ejdb2 (2.0.12) testing; urgency=medium

  * Upgraded to iowow_1.3.16 with critical fixes

 -- Anton Adamansky <adamansky@gmail.com>  Fri, 03 May 2019 11:59:56 +0700

ejdb2 (2.0.11) testing; urgency=medium

  * Upgraded to iowow_1.3.15 with critical fixes

 -- Anton Adamansky <adamansky@gmail.com>  Wed, 01 May 2019 23:34:20 +0700

ejdb2 (2.0.10) testing; urgency=medium

  * CRITICAL: Fixed unexpected database file truncation and data loss on close.
  * Upgraded to iowow 1.3.14

 -- Anton Adamansky <adamansky@gmail.com>  Sat, 27 Apr 2019 01:12:15 +0700

ejdb2 (2.0.9) testing; urgency=medium

  * Upgraded to iowow 1.3.13
  * Building Native JNI binding on Android
  * ejdb2_dart: Added notFound,invalidQuery to EJDB2Error

 -- Anton Adamansky <adamansky@gmail.com>  Thu, 25 Apr 2019 17:47:16 +0700

ejdb2 (2.0.8) testing; urgency=medium

  * Remove redundant memory coping in ejdb_put()

 -- Anton Adamansky <adamansky@gmail.com>  Sun, 21 Apr 2019 14:38:08 +0700

ejdb2 (2.0.7) testing; urgency=medium

  * Fixed deadlock condition with OSX pthread rwlocks.
  * Phreads `PTHREAD_RWLOCK_PREFER_WRITER_NONRECURSIVE_NP` is used by default

 -- Anton Adamansky <adamansky@gmail.com>  Sun, 21 Apr 2019 12:20:27 +0700

ejdb2 (2.0.6) testing; urgency=medium

  * Java JNI binding implemented (#243)

src/bindings/ejdb2_jni/src/main/java/com/softmotions/ejdb2/EJDB2.java
package com.softmotions.ejdb2;

import java.io.ByteArrayOutputStream;
import java.io.OutputStream;
import java.nio.charset.StandardCharsets;

/**
 * EJDB2 JNI Wrapper.
 * <p>
 * Database opened by {@link EJDB2Builder#open()} helper.
 * <p>
 * Example:
 *
 * <pre>
 * {@code
 *  try (EJDB2 db = new EJDB2Builder("my.db").open() {
 *    ...
 *  }
 * }
 * </pre>
 * <p>
 * In order to release memory resources and avoiding data lost every opened
 * database instance should be closed with {@link EJDB2#close()}.
 *
 * @author Adamansky Anton (adamansky@softmotions.com)
 */
public final class EJDB2 implements AutoCloseable {

  static {
    System.loadLibrary("ejdb2jni");
  }

  @SuppressWarnings("unused")
  private long _handle;

  private final EJDB2Builder options;

  /**
   * Returns options used to build this database instance.
   *
   * @return
   */
  public EJDB2Builder getOptions() {
    return options;
  }

  EJDB2(EJDB2Builder options) throws EJDB2Exception {
    this.options = options;
    _open(options);
  }

  /**
   * Closes database instance and releases all resources.
   */
  @Override
  public void close() {
    _dispose();
  }

  /**
   * Create a query instance. Query can be reused multiple times with various
   * placeholder parameters. See JQL specification:
   * https://github.com/Softmotions/ejdb/blob/master/README.md#jql
   * <p>
   * Note: collection name must be encoded in query, eg: {@code @mycoll/[foo=bar]}
   * </p>
   */
  public JQL createQuery(String query) throws EJDB2Exception {
    return new JQL(this, query, null);
  }

  /**
   * Create a query instance. Query can be reused multiple times with various
   * placeholder parameters. See JQL specification:
   * https://github.com/Softmotions/ejdb/blob/master/README.md#jql
   * <p>
   * If {@code collection} is not null it will be used for query. In this case
   * s   * collection name encoded in query will not be taken into account.
   *
   * @param collection Optional collection name
   */
  public JQL createQuery(String query, String collection) throws EJDB2Exception {
    return new JQL(this, query, collection);
  }

  /**
   * Persists {@code json} document into {@code collection}.
   *
   * @param collection Collection name
   * @param json       JSON document
   * @return Generated identifier for document
   * @throws EJDB2Exception
   */
  public long put(String collection, String json) throws EJDB2Exception {
    return _put(collection, json, 0);
  }

  /**
   * Persists {@code json} document under specified {@code id}.
   * <p>
   * If {@code id} is zero a new document identifier will be generated.
   *
   * @param collection Collection name
   * @param json       JSON document
   * @param id         Document id. If zero a new identifier will be genareted
   * @return Document identifier
   * @throws EJDB2Exception
   */
  public long put(String collection, String json, long id) throws EJDB2Exception {
    return _put(collection, json, id);
  }

  /**
   * Removes a document identified by given {@code id} from collection
   * {@code coll}.
   * <p>
   * If document is not found {@link EJDB2Exception} will be thrown:
   *
   * <pre>
   * {@code
   *  code: 75001
   *  message: Key not found. (IWKV_ERROR_NOTFOUND)
   * }
   * </pre>
   *
   * @param collection Collection name
   * @param id         Document id
   * @throws EJDB2Exception
   */
  public void del(String collection, long id) throws EJDB2Exception {
    _del(collection, id);
  }

  public void renameCollection(String oldCollectionName, String newCollectionName) throws EJDB2Exception {
    _rename_collection(oldCollectionName, newCollectionName);
  }

  /**
   * Apply rfc6902/rfc7386 JSON patch to the document identified by {@code id}.
   *
   * @param collection Collection name
   * @param patch      JSON patch
   * @param id         Document id
   * @throws EJDB2Exception
   */
  public void patch(String collection, String patch, long id) throws EJDB2Exception {
    _patch(collection, patch, id, false);
  }

  /**
   * Apply JSON merge patch (rfc7396) to the document identified by `id` or
   * insert new document under specified `id`.
   */
  public void patchOrPut(String collection, String patch, long id) {
    _patch(collection, patch, id, true);
  }

  /**
   * Write document identified by {@code id} into {@code out}.
   * <p>
   * If document is not found {@link EJDB2Exception} will be thrown:
   *
   * <pre>
   * {@code
   *  code: 75001
   *  message: Key not found. (IWKV_ERROR_NOTFOUND)
   * }
   * </pre>
   *
   * @param collection
   * @param id
   * @param out
   */
  public void get(String collection, long id, OutputStream out) {
    _get(collection, id, out, false);
  }

  /**
   * Returns document identified by {@code id} as String.
   * <p>
   * If document is not found {@link EJDB2Exception} will be thrown:
   *
   * <pre>
   * {@code
   *  code: 75001
   *  message: Key not found. (IWKV_ERROR_NOTFOUND)
   * }
   * </pre>
   *
   * @param collection
   * @param id
   */
  public String getAsString(String collection, long id) {
    ByteArrayOutputStream bos = new ByteArrayOutputStream();
    get(collection, id, bos);
    return bos.toString(StandardCharsets.UTF_8);
  }

  /**
   * Same as {@link #get(String, long, OutputStream)} but returns prettified JSON
   * document.
   *
   * @param collection Collection name
   * @param id         Document id
   * @param out        Output stream to write
   */
  public void getPrettified(String collection, long id, OutputStream out) {
    _get(collection, id, out, true);
  }

  /**
   * Returns JSON document describind database structure.
   * <p>
   * Example database metadata:
   *
   * <pre>
   * {@code
   *   {
   *    "version": "2.0.0", // EJDB engine version
   *    "file": "db.jb",    // Path to storage file
   *    "size": 16384,      // Storage file size in bytes
   *    "collections": [    // List of collections
   *     {
   *      "name": "c1",     // Collection name
   *      "dbid": 3,        // Collection database ID
   *      "rnum": 2,        // Number of documents in collection
   *      "indexes": [      // List of collections indexes
   *       {
   *        "ptr": "/n",    // rfc6901 JSON pointer to indexed field
   *        "mode": 8,      // Index mode. Here is EJDB_IDX_I64
   *        "idbf": 96,     // Index database flags. See iwdb_flags_t
   *        "dbid": 4,      // Index database ID
   *        "rnum": 2       // Number records stored in index database
   *       }
   *      ]
   *     }
   *    ]
   *   }
   * }
   * </pre>
   *
   * @param out Output stream to write
   * @throws EJDB2Exception
   */
  public void info(OutputStream out) throws EJDB2Exception {
    _info(out);
  }

  /**
   * Returns JSON document describind database structure. Same as
   * {@link #info(OutputStream)} but returns JSON data as string.
   *
   * @throws EJDB2Exception
   */
  public String infoAsString() throws EJDB2Exception {
    ByteArrayOutputStream bos = new ByteArrayOutputStream();
    info(bos);
    return bos.toString(StandardCharsets.UTF_8);
  }

  /**
   * Removes {@code collection} from database and all its documents.
   *
   * @param collection Collection name
   * @return Current database instance
   * @throws EJDB2Exception
   */
  public EJDB2 removeCollection(String collection) throws EJDB2Exception {
    _remove_collection(collection);
    return this;
  }

  /**
   * Create string index with specified parameters if it has not existed before.
   * <p>
   * Where {@code path} must be fully specified as rfc6901 JSON pointer.
   * <p>
   * Example document:
   *
   * <pre>
   * {@code
   *    "address" : {
   *      "street": "High Street"
   *    }
   * }
   * </pre>
   * <p>
   * Call {@code ensureStringIndex("mycoll", "/address/street", true)} in order to
   * create unique index over all street names in nested address object.
   *
   * @param collection Collection name
   * @param path       JSON pointer path to indexed field
   * @param unique     If {@code true} an unique index will be created
   * @return Current database instance
   * @throws EJDB2Exception
   */
  public EJDB2 ensureStringIndex(String collection, String path, boolean unique) throws EJDB2Exception {
    _ensure_index(collection, path, 0x04 | (unique ? 0x01 : 0));
    return this;
  }

  /**
   * Removes collection index for JSON document field pointed by {@code path}
   *
   * @param collection Collection name
   * @param path       JSON pointer path to indexed field
   * @param unique     {@code true} for unique index
   * @return
   * @throws EJDB2Exception
   */
  public EJDB2 removeStringIndex(String collection, String path, boolean unique) throws EJDB2Exception {
    _remove_index(collection, path, 0x04 | (unique ? 0x01 : 0));
    return this;
  }

  /**
   * Create integer index with specified parameters if it has not existed before.
   *
   * @param collection Collection name
   * @param path       JSON pointer path to indexed field
   * @param unique     {@code true} for unique index
   * @return
   * @throws EJDB2Exception
   */
  public EJDB2 ensureIntIndex(String collection, String path, boolean unique) throws EJDB2Exception {
    _ensure_index(collection, path, 0x08 | (unique ? 0x01 : 0));
    return this;
  }

  /**
   * Removes collection index for JSON document field pointed by {@code path}
   *
   * @param collection Collection name
   * @param path       JSON pointer path to indexed field
   * @param unique     {@code true} for unique index
   * @return
   * @throws EJDB2Exception
   */
  public EJDB2 removeIntIndex(String collection, String path, boolean unique) throws EJDB2Exception {
    _remove_index(collection, path, 0x08 | (unique ? 0x01 : 0));
    return this;
  }

  /**
   * Create floating point number index with specified parameters if it has not
   * existed before.
   *
   * @param collection Collection name
   * @param path       JSON pointer path to indexed field
   * @param unique     {@code true} for unique index
   * @return
   * @throws EJDB2Exception
   */
  public EJDB2 ensureFloatIndex(String collection, String path, boolean unique) throws EJDB2Exception {
    _ensure_index(collection, path, 0x10 | (unique ? 0x01 : 0));
    return this;
  }

  public EJDB2 removeFloatIndex(String collection, String path, boolean unique) throws EJDB2Exception {
    _remove_index(collection, path, 0x10 | (unique ? 0x01 : 0));
    return this;
  }

  /**
   * Creates an online database backup image and copies it into the specified `targetFile`.
   * During online backup phase read/write database operations are allowed and not
   * blocked for significant amount of time. Returns backup finish time as number
   * of milliseconds since epoch.
   *
   * Online backup guaranties what all records before finish timestamp will
   * be stored in backup image. Later, online backup image can be
   * opened as ordinary database file.
   *
   * @param targetFile Backup file path
   * @note In order to avoid deadlocks: close all opened database cursors
   * before calling this method or do call in separate thread.
   */
  public long onlineBackup(String targetFile) throws EJDB2Exception {
    return _online_backup(targetFile);
  }

  private native void _open(EJDB2Builder opts) throws EJDB2Exception;

  private native void _dispose() throws EJDB2Exception;

  private native long _put(String collection, String json, long id) throws EJDB2Exception;

  private native void _del(String collection, long id) throws EJDB2Exception;

  private native void _rename_collection(String oldCollectionName, String newCollectionName) throws EJDB2Exception;

  private native void _patch(String collection, String patch, long id, boolean upsert) throws EJDB2Exception;

  private native void _get(String collection, long id, OutputStream out, boolean pretty) throws EJDB2Exception;

  private native void _info(OutputStream out) throws EJDB2Exception;

  private native void _remove_collection(String collection) throws EJDB2Exception;

  private native void _ensure_index(String collection, String path, int mode) throws EJDB2Exception;

  private native void _remove_index(String collection, String path, int mode) throws EJDB2Exception;

  private native long _online_backup(String targetFile) throws EJDB2Exception;
}
